conans/test/unittests/model/editable_cpp_info/apply_test.py

```diff
@@ -25,63 +25,49 @@
     #[{namespace}bindirs]
     #{path_prefix}dirs/bindirs
 
-    [{namespace}extradirs]
-    {path_prefix}dirs/extradirs
-    """))
+    """)
 
 
 class ApplyEditableCppInfoTest(unittest.TestCase):
-    content = '\n\n'.join([
-        base_content.format(namespace="", path_prefix=""),
-        base_content.format(namespace="*:", path_prefix="all/"),
-        base_content.format(namespace="libA:", path_prefix="libA/")
-    ])
 
     def setUp(self):
         self.test_folder = temp_folder()
         self.layout_filepath = os.path.join(self.test_folder, "layout")
-        save(self.layout_filepath, self.content)
+
+    def tearDown(self):
+        shutil.rmtree(self.test_folder)
 
     def test_require_no_namespace(self):
-        editable_cpp_info = EditableCppInfo.load(self.layout_filepath, require_namespace=False)
-        cpp_info = namedtuple('_', EditableCppInfo.cpp_info_dirs)
-        editable_cpp_info.apply_to('libA', cpp_info, base_path=None, settings=None, options=None)
-        self.assertTrue(editable_cpp_info.has_info_for('any-thing'))
-        self.assertListEqual(cpp_info.includedirs, [os.path.normpath('dirs/includedirs'), ])
-        self.assertListEqual(cpp_info.libdirs, [os.path.normpath('dirs/libdirs'), ])
-        self.assertListEqual(cpp_info.resdirs, [os.path.normpath('dirs/resdirs'), ])
-        self.assertListEqual(cpp_info.bindirs, [])
+        content = base_content.format(namespace="", path_prefix="")
+        save(self.layout_filepath, content)
+        editable_cpp_info = EditableCppInfo.load(self.layout_filepath, allow_package_name=False)
+        cpp_info = CppInfo(None)
+        editable_cpp_info.apply_to('libA', cpp_info, settings=None, options=None)
+        self.assertListEqual(cpp_info.includedirs, ['dirs/includedirs'])
+        self.assertListEqual(cpp_info.libdirs, ['dirs/libdirs'])
+        self.assertListEqual(cpp_info.resdirs, ['dirs/resdirs'])
+        # The default defined by package_info() is maintained
+        self.assertListEqual(cpp_info.bindirs, ['bin'])
```

Comment:
I think it shouldn't be maintained. The user won't expect his/her layout to be mixed with any other thing.

Reply:
I think it should be respected, specially when users are requesting to be able to define their layout also in the ``package_info()``. If a package_info() defines a layout that is valid for edition wrt the library paths, but then the editable package only needs to map, for example the ``includedirs``, then, you loose all the logic already implemented in ``package_info()``. Lets discuss with @lasote

Reply:
Mhh and think I'm ok keeping the cpp_info as a default, but I don't understand why `dirs/bindirs` is not there.

Reply:
Oh! now I see. Looks good for me.

Reply:
I understand the reasons.

I changed it because when I was using the feature I really expected the opposite. Why Conan would remove my ``includedirs`` that is already pointing to the ``include`` folder which is always valid for my layout (either cache or editable), and will remove it, and will force me to repeat myself, adding it to the layout file (and to every layout file for every configuration Windows, Linux, etc).

But I agree this is totally a matter of UX, and no problem to discuss, and I am fine with either approach, as long as we agree explicitly on that behavior.

Reply:
Please @uilianries @SSE4 your feedback (as heavy conan users) on what would you expect from this would be very valuable. In short, these are the 2 options:

A) Once a layout file is defined, it will cancel all the ``xxxdirs`` that are defined by ``package_info()``
B) Only what is defined in the layout file overwrites what is defined by ``package_info()``

Reply:
I also like approach A, as it's more explicit, but for me it appears like we have an inconsistency between `package_info` and `layout` abilities.
for instance, with `package_info`, I may define per-configuration layouts, like:
```
self.cpp_info.defines = ["ALL_CONFIGS"]
self.cpp_info.debug.defines = ["DEBUG"]
self.cpp_info.release.defines = ["NDEBUG"]
```
however, it doesn't appear like I can achieve the same effect with layouts. or am I missing something?
if it's the case, probably users will still want to be able to define per-configuration layouts.

Reply:
Because defines, cppflags, etc, they don't change. Only the "layout" of the folders, need to be re-mapped. Same with library names the library names will be mostly reused from ``package_info()``.

Reply:
yeah, but what about this:
```
self.cpp_info.libdirs= ["libs_common"]
self.cpp_info.debug.libdirs= ["libs_debug"]
self.cpp_info.release.libdirs= ["libs_release"]
```
can this be defined via layouts?

Reply:
and the next question then, will definition of `libdirs` in layout invalidate `debug.libdirs` and `release.libdirs` from package info?

Reply:
yeah, that's my point - users will probably expect feature parity between `layout` and `package_info`, otherwise, layout will override/invalidate things that it cannot handle without providing an easy way to avoid it

Reply:
That is the point of my proposal:
- It is very likely that in most cases, the normal compilation information defined in ``package_info()``, like definitions, flags, library names, etc. won't be, and probably shouldn't be re-defined by a layout file
- The layout files only affect by definition to paths: ``includedirs`` and family
- Then, the final result will always be a mixture of 2 sources of information: the ``package_info()`` method, and the ``layout`` file.
- If we don't do this, there will be immediate complains about users that need to repeat themselves.
- Note that the layout files do not contain any logic, it is necessary to switch and use a different file if you want to change the layout. It is typical that some definitions are conditional, like those BOOST_USE_DYNAMIC-like when using shared libraries, so it would be a nightmare to have to define those in layout files.
- Adding logic to layouts was done in conan workspaces, with ``eval()``. While doable, it is a new syntax that users need to learn. Maybe we want to go this way instead of switching between layout files?

Reply:
I am not sure, to be honest. e.g. how does switching layout matches with typical developer's workflow, e.g. someone who uses IDE like Visual Studio, CLion, XCode, etc. and tries to switch between build configurations in his favorite IDE.
now developer also has to remember that he has to switch layout each time he switches configuration. or probably, somehow to do it automatically (but how?).
okay, now let's switch to the command line use-case, where developer just uses bunch of conan commands, or may be even conan package tools, how now he should always keep in mind to switch layouts in time? should we add arguments to the existing commands to accept layout, so they call `conan link` internally (just like in case of `--profile` argument)? should we make CPT aware about this new feature?

Reply:
IMO there are two different things:
 
* There are _placeholders_ available in the layout files, you can write something like `mydir/{settings.build_type}/folder` and it will be resolved to `mydir/Debug/folder` or `mydir/Release/folder` (same with options and the other settings). I'm sure this is not a solution, and won't fit user needs, but it declares a path to walk. If we need conditional logic inside the layout files I will go for a template system instead of python `eval`, I find it more powerful and maintainable.

 * A completely different thing is being able to specify a file when linking a package as editable. I find it quite useful, but IMO it is not for linking debug-release (previous point) or linux-win, but to use a file for a library, a different file for another library,... and let the user define those files. Up to now we can just use the file inside the repo or the `layouts/default` one, which is not fair.

Reply:
I have added a new functional test to validate the placeholders approach might be valid for the Debug/Release case.

The idea is that the different layout files are to differentiate between different OS or build systems, while the placeholders should be good to handle different configurations within the same OS and build system, i.e. the different Debug/Release, x86/64, etc. typical variants.